Autodesk: Update the LineStyle proposal from suggestions

erikaharrison-adsk opened the pull request:
### Description of Proposal

It is preferred to use a separate schema for the DashDotLines, because the dash dot patterns can only be applied for the lines. If we directly add style property to the BasisCurves, the style property will be only valid when the type is linear. This constraint is not convenient. We also add a separate rprim for the new primitive. In this proposal, I add a separate section talking about the extents of the DashDotLines, and a section talking about how to handle screen space pattern. At the end of the proposal, two examples are provided.

<!--Replace the link here with a link to your own proposal markdown file -->
[Link to Rendered Proposal](https://github.com/autodesk-forks/USD-proposals/blob/adsk/feature/LineStyle/proposals/_notPublished/hold/LineStyle/README.md)

### Supporting Materials

Previous PRs:

* https://github.com/PixarAnimationStudios/OpenUSD-proposals/pull/47
* https://github.com/PixarAnimationStudios/OpenUSD-proposals/pull/27
* https://github.com/PixarAnimationStudios/OpenUSD-proposals/pull/20

### Contributing

<!--
Please review the  [Contributing](https://graphics.pixar.com/usd/release/contributing_to_usd.html) page in the
documentation for the Supplemental Terms that apply to this repository.
Place an X in the box when you have reviewed and agree to the Supplemental Terms.
-->
- [X] I agree to and accept the [Supplemental Terms](https://graphics.pixar.com/usd/release/contributing_supplemental.html).

PierreWang committed:
Update the LineStyle proposal according to the suggestions in the PR.
It is preferred to use a separate schema for the DashDotLines, because the dash dot patterns can only be applied for the lines. If we directly add style property to the BasisCurves, the style property will be only valid when the type is linear. This constraint is not convenient.
We also add a separate rprim for the new primitive.
In this proposal, I add a separate section talking about the extents of the DashDotLines, and a section talking about how to handle screen space pattern.
At the end of the proposal, two examples are provided.

jcowles commented on proposals/LineStyle/README.md:
```diff
@@ -1,54 +1,66 @@
+- widths. Widths are now interpreted as the widths in screen space.
+
+The DashDotLines has the following new properties:
+- screenSpacePattern. A bool uniform. By default it is true, which means the dash-dot pattern will be based on screen unit. If we zoom in, the pattern on the line will change in the world space, so that the dash size and the dash gap size in the screen space will not change. If it is false, the pattern will be based on world unit. If we zoom in, the pattern on the line will not change in world sp ...
```
I think "screenSpace" should be one word, "screenspace"

PierreWang replied:
Yes, you are right.

jcowles commented on proposals/LineStyle/README.md:
```diff
@@ -1,54 +1,66 @@
 
-If the curve style is "sketch", "dashDot" or "screenSpaceDashDot", the curve must bind to a specific material. The property of the style, such as the cap shape or the scale of the dash dot pattern, will be set in the material via the surface input.
+### Extents of the DashDotLines
+Different from the other Curves, the extents of the DashDotLines is only the bound box of the control points. The width of the line will not be considered, because it is screen spaced, that it is implemented via the shader.
```
This is conditional on the screenspacePattern attribute - it seems when screenspacePattern is false, then the line does have an extent in worldspace, no?

On the other hand, if screenspacePattern should /only/ be considered for rendering, and DashDotLines have no extent, then perhaps it should exclusively be on the shader/material or be a primvar (like start/endCapType, patternScale)?

PierreWang replied:
It is not conditional on the screenspacePattern attibute.
"The width of the line will not be considered, because it is screen spaced, that it is implemented via the shader." In this sentence I mean the width is screenspaced. The attribute screenspacePattern only control if the pattern is screenspaced. But no matter whether it is true or false, the width of the line will always be screenspaced.
So DashDotLines will always have an extents, which is the bound box of the control points.

jcowles reviewed (commented):
Here are a couple comments, mostly just noodling the details.

The bigger question in my mind is still if this schema is needed - I totally understand the utility of it, but it's also yet another schema to support everywhere, in addition to the underlying existing curves...

PierreWang commented:
> The bigger question in my mind is still if this schema is needed - I totally understand the utility of it, but it's also yet another schema to support everywhere, in addition to the underlying existing curves...

As @nvmkuruc pointed, the dash-dot patterned lines will interpret the width as screen-space width. This is different from the definition of a common Curve. To avoid schema slicing, it is better to use a separate typed schema.

PierreWang committed:
Update the Readme.
We will use a separate Typed schema DashDotPattern to save the pattern, instead to use a texture to save the pattern. A DashDotLines primitive will not bind a material. Instead, it can bind a DashDotPattern.

PierreWang committed:
Change the uppercase letter "S" in "screenSpace" to lowercase.

PierreWang commented:
In the new change, we create a new schema DashDotPattern. The pattern can be set via an array of float2. So we don't need to use a texture to provide the pattern. And we don't require that the DashDotLines primitive must binds a material. Instead, if the DashDotLines primitive has a pattern, it should bind a DashDotPattern.

spiffmon commented on proposals/LineStyle/README.md:
```diff
@@ -1,76 +1,140 @@
+Different from the other Curves, the extents of the DashDotLines is only the bound box of the control points. The width of the line will not be considered, because it is screen spaced, that it is implemented via the shader.
+
+### The DashDotPattern schema
+The DashDotPattern schema inherits from Typed. It saves the detail of a dash-dot pattern. A dash-dot pattern is an array of symbols. Each symbol is either a dash, which is a line, or a dot, which is a point. The DashDotPattern can be bound to one or more DashDotLines primitive.
+
+The DashDotPattern has the following new properties:
+- patternPeriod. A float uniform. It is the length of a pattern. By default it is zero. If there is no pattern, it should be zero.
+- pattern. An array of float2. It saves the dash-dot pattern. For each float2, the x value and the y value must be zero or positive. The x value saves the offset of the start of current symbol, from the end of the previous symbol. If the current symbol is the first symbol, the offset is from the start of the pattern to the start of current symbol. The y value saves the length of the current symbo ...
+        
+For example, assume the pattern is [(0, 10), (1, 4), (3, 0)]. It means the first symbol is a dash which is from 0 to 10. The second symbol is a dash which is from 11 to 15, and the third symbol is a dot which is at position 18. There are gaps between 10 and 11, and between 15 and 18. If the patternPeriod is 20, there is also a gap between 18 and 20.
+
+An API schema DashDotPatternAPI is provided to bind the DashDotPattern to a DashDotLines primitive.
```
Specifying the Pattern as a separate prim that must be bound via relationship seems like overkill here, and imposes more complexity/cost on the scenegraph and Hydra efficiency than a simpler, API-only solution would.  UsdShadeMaterial must be bound because it consists of multiple prims, and because the binding mechanism itself needs to be very "rich" (i.e. complicated).

As an alternative, could we get rid of the `DashDotPattern` Typed schema, instead putting all of its properties into `DashDotPatternAPI`?  Then, the simple case is _very_ simple: all the information for a single DashDotLines is contained **on the prim**, very efficiently processed by Hydra.  In the mentioned case of wanting to "bind" the same Pattern onto multiple `DashDotLines` primitives, this seems like a CSS-inspired dream case for the use of class inheritance in USD.  

Now we can't build inheritance into the schema itself (composition doesn't consult the `UsdSchemaRegistry`), but we _can_ provide "strong guidance" in the spec (and helper functions in the schema classes) that `DashDotLines` primitives should minimally inherit from a "_class_DashDotPattern", and potentially also more specific/granular classes tailored to the asset/scene.  So, each of the `DashDotLine` primitives in the "Diagram1" asset might look something like:

```
def DashDotLines "TopArc" (
    prepend inherits = [</_class_Diagram1_DD_Pattern>, </_class_DashDotPattern>]
)
{
}
```

so then anyone referencing that asset could affect all of its `DashDotLines` (and no others) by putting opinions for `DashDotPatternAPI` in a spec for `</_class_Diagram1_DD_Pattern> `.

PierreWang replied:
Thank you. I will modify the proposal. And also for the text proposal, I will also merge the TextStyle into TextStyleAPI, and the same for ParagraphStyle and ColumnStyle.

PierreWang committed:
Merge the DashDotPattern into DashDotPatternAPI schema

PierreWang commented:
Hi @spiffmon , I have updated the proposal. The properties of DashDotPattern is now in DashDotPatternAPI. I don't know if my sample is right:
`def DashDotLines "StyledPolyline" (
    inherits = [</Pattern>]
){
    uniform bool screenspacePattern = true
    uniform token startCapType = "triangle"
    uniform token endCapType = "triangle"
    uniform float patternScale = 11
    int[] curveVertexCounts = [3, 4]
    point3f[] points = [(0, 0, 0), (10, 10, 0), (10, 20, 0), (0, 30, 0), (-10, 40, 0), (-10, 50, 0), (0, 60, 0)]
    float[] widths = [10] (interpolation = "constant")
    color3f[] primvars:displayColor = [(0, 0, 1)]
}
def "Pattern" (
    prepend apiSchemas = ["DashDotPatternAPI"]
)
{
    uniform float patternPeriod = 10
    uniform float2[] pattern    = [(0, 5), (2.5, 0)]
}`
In practice, I found that I could **not** read the patternPeriod and pattern property from the "StyledPolyline" prim in the adapter of DashDotLines. I think as I used inheritance, the properties should be available in the prim.

spiffmon commented:
Yes, `inherits` should make the API schema be present on `</StyledPolyline>`, and you should be able to confirm that in usdview or any python shell.  I don't really know what might be the problem in your Hydra adapter, though I know that in Hydra 2 the general pattern would be to have a SI specifically for `DashDotPatternAPI`... others here might be able to help more, if you can provide a bit more context on the code?

PierreWang commented:
> Yes, `inherits` should make the API schema be present on `</StyledPolyline>`, and you should be able to confirm that in usdview or any python shell. I don't really know what might be the problem in your Hydra adapter, though I know that in Hydra 2 the general pattern would be to have a SI specifically for `DashDotPatternAPI`... others here might be able to help more, if you can provide a bit more context on the code?

Thank you. I made a further check today and found a bug in my code. There is no problem now.

PierreWang commented:
The implementation of this proposal is here:
https://github.com/PixarAnimationStudios/OpenUSD/pull/3151

PierreWang committed:
Modify the document according to latest implementation.

PierreWang committed:
Merge remote-tracking branch 'origin/main' into adsk/feature/LineStyle

PierreWang commented:
It's a long time from my last update, because I am working on other features. 
Now I made some updates to the proposal, adding the shapeDetail and jointType. 
"shapeDetail" is added, because we found that many applications just need a very simple implementation for the DashDotLines, especially when it is only used as the style of the selection rectangle. They can just use the line weight provided by the graphics API, and they don't need the caps and joint. This token will indicate if the application wants a simple implementation or a complete implementation.
"jointType" is added for different jointTypes. Currently only "round" is supported. But the user could provide extended support such as "bevel", or other types.

PierreWang committed:
Move the pictures in proposals/LineStyle to proposals/_notPublished/hold/LineStyle, because the file structure is changed.
